ta/os_test/Android.mk

```diff
@@ -1,4 +1,7 @@
 LOCAL_PATH := $(call my-dir)
 
 local_module := 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b.ta
+ifeq ($(CFG_TA_DYNLINK),y)
+local_dependency_ta := ffd2bded-ab7d-4988-95ee-e4962fff7154.ta
```

Comment:
That is not correct. The true dependency is: `<out-dir-for-os_test>/$(local_module).elf: <out-dir-for-os_test_lib>/libos_test.so`. Not describing the exact dependencies might result in build errors.

Reply:
is it better to make the compiling of libos_test.so a separate module, and set the os_test ta and os_test_lib ta and other tas to depends on libos_test.so?

Now it seems that the os_test ta depends on the os_test_lib ta, and that os_test_lib ta depends on the libos_test.so, so making  os_test ta depends on the os_test_lib ta finally meets the requirement of os_test ta .

BTW, two small questions here:
1. is libos_test.so need to be integrated into the device as the ta files?
2. not sure if my understanding above is right, if without libos_test.so the compiling of os_test ta will fail, then it might be better to not add the check of ifeq ($(CFG_TA_DYNLINK),y). If libos_test.so is only required when  CFG_TA_DYNLINK is enabled, then the check is necessary here.

Reply:
> [...] so making os_test ta depends on the os_test_lib ta finally meets the requirement of os_test ta 

No, it is not equivalent. Try this: `make; rm <out-path>/libos_test.so; touch <src-path>/ta/os_test_lib/os_test_lib.c; make`. It will fail (unless you somehow have added `libos_test.so` to the 'all' target, which should not be the case, see below).

> 1. is libos_test.so need to be integrated into the device as the ta files?

No, it is needed only at build (link) time. The `.ta` file is a signed version of the `.so`, OP-TEE OS can only load signed files.

> 2. [...] it might be better to not add the check of ifeq ($(CFG_TA_DYNLINK),y)

Well, the library (`ffd2bded-ab7d-4988-95ee-e4962fff7154.ta`) is not used if `CFG_TA_DYNLINK != y`, so it seems more logical not to build it at all.

Reply:
I'll try to make the compiling of libos_test.so a separate module.

Reply:
@jforissier Thinking more about this, maybe the fixup patch isn't needed. Similar to how `ta/Makefile` specifies `os_test_lib` before `os_test` in `TA_DIRS` to get the build order, and hence dependency, right, specifying `local_dependency_ta := ffd2bded-ab7d-4988-95ee-e4962fff7154.ta` for `local_module := 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b.ta` is really just to get the AOSP build system to call `os_test_lib/Makefile` before `os_test/Makefile`. It isn't really specifying a Makefile target dependency in the strict sense of the example above (`<out-dir-for-os_test>/$(local_module).elf: <out-dir-for-os_test_lib>/libos_test.so`), i.e. we're not really specifying `5b9e0e40-2636-11e1-ad9e-0002a5d5c51b.ta: ffd2bded-ab7d-4988-95ee-e4962fff7154.ta`.

Reply:
Hmmm, you may very well be right :astonished: 

I completely missed the fact that AOSP runs `$(MAKE)` for each TA or library directory. I was incorrectly thinking that the AOSP build was much more different, that all the rules were imported in a single Makefile (like what happens in optee_os).

Since the `$(MAKE)` line https://github.com/OP-TEE/optee_os/blob/master/mk/aosp_optee.mk#L94 does not specify any target, it will run `all:` and thus we should get all our files (`.so`, `.ta`), indeed.

Sorry for the bad advice :(

Reply:
Yep, np. Your advice/comments are always appreciated! ;) Will remove fixup and add tag. Thanks!